src/core/layertree/qgslayertreemodel.cpp

```diff
@@ -1299,10 +1303,41 @@ QList<QgsLayerTreeModelLegendNode*> QgsLayerTreeModel::layerLegendNodes( QgsLaye
 
 void QgsLayerTreeModel::legendInvalidateMapBasedData()
 {
+  // we have varying icon sizes, and we want icon to be centered and
+  // text to be left aligned, so we have to compute the max width of icons
+  //
+  // we do that for nodes who share a common parent
+  //
+  // we do that here because for symbols with size defined in map units
+  // the symbol sizes changes depends on the zoom level
+
+  QList<QgsSymbolV2LegendNode*> symbolNodes;
+  QMap<QString, int> widthMax;
   foreach ( const LayerLegendData& data, mLegend )
   {
     foreach ( QgsLayerTreeModelLegendNode* legendNode, data.originalNodes )
+    {
       legendNode->invalidateMapBasedData();
+      QgsSymbolV2LegendNode* n = dynamic_cast<QgsSymbolV2LegendNode*>( legendNode );
+      if ( n )
+      {
+        n->setParent( this ); // map scale are in the model, so the parent needs to be set
+        const QSize sz( n->minimumIconSize() );
+        const QString parentKey( n->data( QgsLayerTreeModelLegendNode::ParentRuleKeyRole ).toString() );
+        widthMax[parentKey] = qMax( sz.width(), widthMax.contains( parentKey ) ? widthMax[parentKey] : 0 );
+        n->setIconSize( sz );
+        symbolNodes.append( n );
+      }
+    }
+  }
+
+  foreach ( QgsSymbolV2LegendNode* n, symbolNodes )
+  {
+    const QString parentKey( n->data( QgsLayerTreeModelLegendNode::ParentRuleKeyRole ).toString() );
+    Q_ASSERT( widthMax[parentKey] > 0 );
+    const int twiceMarginWidth = 2; // a one pixel margin avoids hugly rendering of icon
+    n->setIconSize( QSize( widthMax[parentKey] + twiceMarginWidth, n->iconSize().rheight() + twiceMarginWidth ) );
+    n->invalidateMapBasedData();
```

Comment:
It is apparently, at least it seems necessary to call that last (after set iconSize).